### PR TITLE
CCQ: Server allow list

### PR DESCRIPTION
**Pull request by bruce-riley**
This PR adds a configurable allow list to the `ccq_server` so that we can specify which users are allowed to make queries and which RPC calls they are allowed to make.

### CODE REVIEW

File: node/cmd/ccq/utils.go [hunk 1]
```diff
@@ -39,3 +49,151 @@ func FetchCurrentGuardianSet(rpcUrl, coreAddr string) (*common.GuardianSet, erro
...
+	ContractAddress string `json:"contractAddress"`
+	Call            string `json:"call"`
+}
+
+type Permissions map[string]*permissionEntry
+
+type permissionEntry struct {
+	userName     string
+	apiKey       string
+	allowedCalls allowedCallsForUser // Key is something like "thCall:2:000000000000000000000000b4fbf271143f4fbf7b91a5ded31805e42b2208d6:06fdde03"
```
Comment (kev1n-peters) [inline node/cmd/ccq/utils.go:78]: Should this be `"ethCall"`

File: node/cmd/ccq/devnet.config.json [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+  "permissions": [
+    {
+      "userName": "Test User",
+      "apiKey": "my_secret_key",
+      "allowedCalls": [
+        {
+          "ethCall": {
+            "note:": "Name of WETH on Goerli",
+            "chain": 2,
+            "contractAddress": "B4FBF271143F4FBf7B91A5ded31805e42b2208d6",
```
Comment (panoel) [inline node/cmd/ccq/devnet.config.json:11]: Do the contract addresses need to have leading "0x"?

File: node/cmd/ccq/http.go [hunk 3]
```diff
@@ -38,16 +42,24 @@ func (s *httpServer) handleQuery(w http.ResponseWriter, r *http.Request) {
 		return
 	}
 
+	// The API key should be in the header.
+	apiKey, exists := r.Header["X-Api-Key"]
+	if !exists || len(apiKey) != 1 {
```
Comment (panoel) [inline node/cmd/ccq/http.go:47]: The length of the api key is exactly 1?

File: node/cmd/ccq/utils.go [hunk 4]
```diff
@@ -39,3 +49,151 @@ func FetchCurrentGuardianSet(rpcUrl, coreAddr string) (*common.GuardianSet, erro
...
+
+type EthCall struct {
+	Chain           int    `json:"chain"`
+	ContractAddress string `json:"contractAddress"`
+	Call            string `json:"call"`
+}
+
+type Permissions map[string]*permissionEntry
+
+type permissionEntry struct {
```
Comment (panoel) [inline node/cmd/ccq/utils.go:75]: Isn't this just a User struct?

File: node/cmd/ccq/utils.go [hunk 5]
```diff
@@ -39,3 +49,151 @@ func FetchCurrentGuardianSet(rpcUrl, coreAddr string) (*common.GuardianSet, erro
...
+	userName     string
+	apiKey       string
+	allowedCalls allowedCallsForUser // Key is something like "ethCall:2:000000000000000000000000b4fbf271143f4fbf7b91a5ded31805e42b2208d6:06fdde03"
+}
+
+type allowedCallsForUser map[string]struct{}
```
Comment (panoel) [inline node/cmd/ccq/utils.go:81]: You can create a type with an empty struct?

File: node/cmd/ccq/devnet.config.json (hunk 2 above)
Comment (bruce-riley) [reply node/cmd/ccq/devnet.config.json:11]: Nope. It handles both ways.

File: node/cmd/ccq/http.go (hunk 3 above)
Comment (bruce-riley) [reply node/cmd/ccq/http.go:47]: That map look up returns a slice, so it is checking to make sure that there is only API key in the header. I updated the comment.

File: node/cmd/ccq/utils.go (hunk 5 above)
Comment (bruce-riley) [reply node/cmd/ccq/utils.go:81]: Yes, this is how you do a set in go.

File: node/cmd/ccq/utils.go (hunk 4 above)
Comment (bruce-riley) [reply node/cmd/ccq/utils.go:75]: Nope, `allowCalls` is different.

**Review by panoel — APPROVED** [review]

**Pull request merged**